<a href="https://colab.research.google.com/github/gudellidilipkumar/python/blob/main/Week_2_2_wed.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Analyze the performance and cost-effectiveness of marketing campaigns across three platforms (Social Media, Email Marketing, Paid Ads) over three quarters (Q1, Q2, Q3). Build matrices for lead generation and cost-per-lead, calculate the total marketing cost, simulate 15% growth for Q4 lead generation, transpose the Q4 projected lead matrix, and recommend which platform should receive additional budget in Q4 based on the analysis.

## Define the lead generation matrix

### Subtask:
Define a 3x3 matrix representing the number of leads generated by each platform (Social Media, Email Marketing, Paid Ads) in each quarter (Q1, Q2, Q3).


**Reasoning**:
Create a NumPy array to represent the lead generation matrix with 3 rows for platforms and 3 columns for quarters, populated with hypothetical data.



In [1]:
import numpy as np

lead_generation_matrix = np.array([
    [1000, 1200, 1500],  # Social Media (Q1, Q2, Q3)
    [800, 950, 1100],    # Email Marketing (Q1, Q2, Q3)
    [1500, 1800, 2000]   # Paid Ads (Q1, Q2, Q3)
])

print(lead_generation_matrix)

[[1000 1200 1500]
 [ 800  950 1100]
 [1500 1800 2000]]


## Define the cost-per-lead matrix

### Subtask:
Define a 3x3 matrix representing the average cost per lead for each platform in each quarter.


**Reasoning**:
Define a 3x3 NumPy array representing the average cost per lead and populate it with hypothetical data as instructed.



In [2]:
cost_per_lead_matrix = np.array([
    [2.5, 2.3, 2.1],  # Social Media (Q1, Q2, Q3)
    [1.8, 1.7, 1.6],  # Email Marketing (Q1, Q2, Q3)
    [3.0, 2.8, 2.7]   # Paid Ads (Q1, Q2, Q3)
])

print(cost_per_lead_matrix)

[[2.5 2.3 2.1]
 [1.8 1.7 1.6]
 [3.  2.8 2.7]]


## Calculate the total marketing cost matrix

### Subtask:
Multiply the Lead Generation Matrix and the Cost-per-Lead Matrix element-wise to get the total marketing cost per platform in each quarter.


**Reasoning**:
Perform element-wise multiplication of the two matrices to calculate the total marketing cost and store the result.



In [3]:
total_marketing_cost_matrix = lead_generation_matrix * cost_per_lead_matrix
print(total_marketing_cost_matrix)

[[2500. 2760. 3150.]
 [1440. 1615. 1760.]
 [4500. 5040. 5400.]]


## Simulate growth for q4

### Subtask:
Increase the values in the Lead Generation Matrix by 15% to simulate expected lead generation for Q4.


**Reasoning**:
Multiply the lead generation matrix by 1.15 to simulate Q4 lead generation and store the result in a new variable, then print the new variable.



In [4]:
q4_projected_lead_matrix = lead_generation_matrix * 1.15
print(q4_projected_lead_matrix)

[[1150.  1380.  1725. ]
 [ 920.  1092.5 1265. ]
 [1725.  2070.  2300. ]]


## Transpose the projected growth matrix

### Subtask:
Transpose the Q4-projected lead matrix to have platforms as rows and quarters as columns.


**Reasoning**:
Transpose the Q4 projected lead matrix using the `.T` attribute and print the result.



In [5]:
q4_projected_lead_matrix_transposed = q4_projected_lead_matrix.T
print(q4_projected_lead_matrix_transposed)

[[1150.   920.  1725. ]
 [1380.  1092.5 2070. ]
 [1725.  1265.  2300. ]]


## Analyze and recommend

### Subtask:
Analyze the results from the previous steps to evaluate the performance and cost-effectiveness of each platform and recommend which platform should receive additional budget in Q4.


**Reasoning**:
Analyze the total marketing cost and projected lead generation matrices, calculate the projected Q4 cost per lead, evaluate platform performance, and recommend a platform for additional budget.



In [6]:
print("Total Marketing Cost Matrix:")
print(total_marketing_cost_matrix)

print("\nQ4 Projected Lead Matrix:")
print(q4_projected_lead_matrix)

# Calculate projected cost per lead for Q4 (assuming cost per lead remains the same as Q3)
q4_projected_cost_per_lead = cost_per_lead_matrix[:, 2]

print("\nProjected Cost per Lead for Q4:")
print(q4_projected_cost_per_lead)

# Calculate total projected cost for Q4
q4_projected_total_cost = q4_projected_lead_matrix[:, 2] * q4_projected_cost_per_lead

print("\nProjected Total Cost for Q4:")
print(q4_projected_total_cost)

# Evaluate performance and recommend
platforms = ['Social Media', 'Email Marketing', 'Paid Ads']

print("\n--- Platform Performance Evaluation and Recommendation ---")

for i in range(len(platforms)):
    print(f"\nPlatform: {platforms[i]}")
    print(f"  Historical Lead Generation (Q1-Q3): {lead_generation_matrix[i]}")
    print(f"  Total Historical Cost (Q1-Q3): {total_marketing_cost_matrix[i].sum():.2f}")
    print(f"  Projected Q4 Lead Generation: {q4_projected_lead_matrix[i, 2]:.2f}")
    print(f"  Projected Q4 Cost per Lead: {q4_projected_cost_per_lead[i]:.2f}")
    print(f"  Projected Q4 Total Cost: {q4_projected_total_cost[i]:.2f}")

# Recommendation based on projected cost-effectiveness (lower cost per lead) and growth potential
# Email Marketing has the lowest projected cost per lead in Q4
recommended_platform = platforms[np.argmin(q4_projected_cost_per_lead)]

print(f"\nRecommendation: Based on the analysis, {recommended_platform} is projected to have the lowest cost per lead in Q4.")
print(f"Considering its historical performance and projected cost-effectiveness, allocating additional budget to {recommended_platform} in Q4 could potentially yield the most leads per dollar spent.")

Total Marketing Cost Matrix:
[[2500. 2760. 3150.]
 [1440. 1615. 1760.]
 [4500. 5040. 5400.]]

Q4 Projected Lead Matrix:
[[1150.  1380.  1725. ]
 [ 920.  1092.5 1265. ]
 [1725.  2070.  2300. ]]

Projected Cost per Lead for Q4:
[2.1 1.6 2.7]

Projected Total Cost for Q4:
[3622.5 2024.  6210. ]

--- Platform Performance Evaluation and Recommendation ---

Platform: Social Media
  Historical Lead Generation (Q1-Q3): [1000 1200 1500]
  Total Historical Cost (Q1-Q3): 8410.00
  Projected Q4 Lead Generation: 1725.00
  Projected Q4 Cost per Lead: 2.10
  Projected Q4 Total Cost: 3622.50

Platform: Email Marketing
  Historical Lead Generation (Q1-Q3): [ 800  950 1100]
  Total Historical Cost (Q1-Q3): 4815.00
  Projected Q4 Lead Generation: 1265.00
  Projected Q4 Cost per Lead: 1.60
  Projected Q4 Total Cost: 2024.00

Platform: Paid Ads
  Historical Lead Generation (Q1-Q3): [1500 1800 2000]
  Total Historical Cost (Q1-Q3): 14940.00
  Projected Q4 Lead Generation: 2300.00
  Projected Q4 Cost per Lea

## Summary:

### Data Analysis Key Findings

*   The total marketing cost for Q1, Q2, and Q3 across Social Media, Email Marketing, and Paid Ads was calculated. For example, in Q3, Social Media cost was \$3150, Email Marketing cost was \$1760, and Paid Ads cost was \$5400.
*   Projected Q4 lead generation, based on a 15% growth simulation from Q3, shows 1725 leads for Social Media, 1265 leads for Email Marketing, and 2300 leads for Paid Ads.
*   Assuming the Q3 cost per lead remains constant for Q4, the projected cost per lead is \$2.10 for Social Media, \$1.60 for Email Marketing, and \$2.70 for Paid Ads.
*   Email Marketing is projected to have the lowest cost per lead (\$1.60) in Q4.

### Insights or Next Steps

*   Based on the projected cost-effectiveness (lowest cost per lead), allocating additional budget to Email Marketing in Q4 is recommended as it is likely to generate the most leads per dollar spent.
*   Further analysis could involve projecting Q4 total costs based on the projected leads and Q3 cost per lead to understand the financial implications of the simulated growth.
